README
Video-FFmpeg version 0.48
========================

** This UNAUTHORIZED VERSION was done by Rene Schickbauer
** to fix some bugs that completly broke this module
**
** See https://rt.cpan.org/Public/Bug/Display.html?id=73132
** for more information

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

libavformat
ExtUtils::PkgConfig

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2009 by Max Vohra <max@seattlenetworks.com>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Video::FFmpeg.

0.42 Wed Dec 23 10:43:03 2009
	- First public release.
0.43 Wed Dec 29 04:43:20 2009
	- Fixed insufficient thread locking around avcodec_open/close() error;
0.44 Wed Jun 4 12:14:20 2010
	- Fixed typo in metatag functions. Thanks Sergiy Borodych.
0.45 Wed Jun 4 03:40:00 2010
	- Added META.yml
	- Fixed versioning info.
0.46 Wed Jun 4 03:40:00 2010
	- Fixed issue with test video. Variable bitrates give variable responses.
0.47 Wed Oct 27 7:27:00 2010
	- Fixed Mac OS X Compilation issue. Thanks Mischa Spiegelmock.
	- Fixed META.yml desription.
	- Lowered required version of libavformat to compile on Debian Lenny.

0.48 Sun Apr 29 19:00:00 2012
	- Fixed some critical bugs, see https://rt.cpan.org/Public/Bug/Display.html?id=73132
	- Makefile.PL does not find -lavformat (fixed)
	- Compile errors in FFmpeg.xs
	- Some Perl code still uses Switch.pm (deprecated) (fixed)
	- Two Tests had floating point problems (fixed)
	- Uploaded UNAUTHORIZED version


Makefile.PL
use 5.010000;
use ExtUtils::MakeMaker;
use ExtUtils::PkgConfig;

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

my $lavf_lib = "-L/usr/lib " . ExtUtils::PkgConfig->libs("libavformat");
my $lavf_inc = ExtUtils::PkgConfig->cflags("libavformat");
die("Installed libavformat version is too low. I require 52.7 or greater") unless (
	ExtUtils::PkgConfig->atleast_version("libavformat", "52.7.0"));

WriteMakefile(
    NAME              => 'Video::FFmpeg',
    VERSION_FROM      => 'lib/Video/FFmpeg.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.010 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Video/FFmpeg.pm', # retrieve abstract from module
       AUTHOR         => 'Rene Schickbauer <rene.schickbauer@magnapowertrain.com>') : ()),
    LIBS              => [$lavf_lib], # e.g., '-lm'
    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
    INC               => '-I. '.$lavf_inc, # e.g., '-I. -I/usr/include/other'
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);


lib/Video/FFmpeg/AVFormat.pm
use 5.010;
package Video::FFmpeg::AVFormat;
use Video::FFmpeg;

our $VERSION = '0.48';

my $i;
sub new {
	$i = Video::FFmpeg::AVFormat::open($_[1]);
}

sub streams {
    my ($self) = @_;
    my @streams;
	for(0 .. $self->nb_streams-1){
		my $stream = $self->get_stream($_);
		given($stream->codec_type){
			when ("video") {
				bless $stream, 'Video::FFmpeg::AVStream::Video';
				push @streams, $stream;
			}
			when ("audio") {
				bless $stream, 'Video::FFmpeg::AVStream::Audio';
				push @streams, $stream;
			}
			when ("subtitle") {
				bless $stream, 'Video::FFmpeg::AVStream::Subtitle';
				push @streams, $stream;
			}
			default {
				push @streams, $stream;
			}
		}
	};
    return @streams;
}
	
sub audio {
    my ($self) = @_;
    my @streams;
    for(0 .. $self->nb_streams-1){
		my $stream = $self->get_stream($_);
		if($stream->codec_type eq "audio"){
			bless $stream, Video::FFmpeg::AVStream::Audio;
			return $stream if(!wantarray);
	        push @streams, $stream;
		};
    };
    return @streams;
};

sub video {
    my ($self) = @_;
    my @streams;
    for(0 .. $self->nb_streams-1){
		my $stream = $self->get_stream($_);
		if($stream->codec_type eq "video"){
			bless $stream, Video::FFmpeg::AVStream::Video;
			return $stream if(!wantarray);
	        push @streams, $stream;
		};
    };
    return @streams;
};

sub subtitle {
    my ($self) = @_;
    my @streams;
    for(0 .. $self->nb_streams-1){
		my $stream = $self->get_stream($_);
		if($stream->codec_type eq "subtitle"){
			bless $stream, Video::FFmpeg::AVStream::Subtitle;
			return $stream if(!wantarray);
	        push @streams, $stream;
		};
    };
    return @streams;
};

1;
__END__

=head1 NAME

Video::FFmpeg::AVFormat - Retrieve video properties using libavformat such as: height width codec fps

=head1 SYNOPSIS

  use Video::FFmpeg;

  my $info = Video::FFmpeg::AVFormat->new($ARGV[0]);

  print "Duration: ",$info->duration,"\n";

  my @video = $info->video;
  print "num video streams: ",$#video+1,"\n";
  my @audio = $info->audio;
  print "num audio streams: ",$#audio+1,"\n";
  my @sub = $info->subtitle;
  print "num sub streams: ",$#sub+1,"\n";

  my @streams = $info->streams;
  for my $id (0 .. $#streams){
    my $stream = $streams[$id];
    print $stream->codec_type,"stream $id\n";
    print "\ttype: ",$stream->codec_type,"\n";
    print "\tcodec: ",$stream->codec,"\n";
    print "\tlanguage: ",$stream->lang,"\n";
    given($stream->codec_type){
      when("video") {
        print "\tfps: ",$stream->fps,"\n";
        print "\tDAR: ",$stream->display_aspect,"\n";
      }
      when("audio") {
        print "\tsample rate: ",$stream->sample_rate,"hz\n";
        print "\taudio language: ",$stream->lang,"\n";
      }
      when("subtitle") {
        print "\tsub codec: ",$stream->codec,"\n";
        print "\tsub language: ",$stream->lang,"\n";
      }
    }
  };

=head1 METHODS

=head2 The Video::FFmpeg::AVFormat class

=head3 Video::FFmpeg::AVFormat->new($file)

AVFormat Constructor

=head3 filename

returns the filename

=head3 duration

duration of the stream, in HH:MM:SS.MS. 

=head3 start_time

position of the first frame of the component, in microseconds

=head3 bit_rate

total stream bitrate in bit/s, 0 if not available. 

=head3 video

if called in scalar context, returns the first Video::FFmpeg::AVStream::Video object. if called in list context, it returns all Video::FFmpeg::AVStream::Video objects

=head3 audio

if called in scalar context, returns the first Video::FFmpeg::AVStream::Audio object. if called in list context, it returns all Video::FFmpeg::AVStream::Audio objects

=head3 subtitles

if called in scalar context, returns the first Video::FFmpeg::AVStream::Subtitle object. if called in list context, it returns all Video::FFmpeg::AVStream::Subtitle objects

=head3 streams

returns a list of all Video::FFmpeg::AVStream objects.


=head2 The Video::FFmpeg::AVStream class

=head3 codec

name of the codec

=head3 codec_type

returns one of "audio", "video", "subtitle", "data", "attachment", "data", or "unknown"

=head3 lang

returns the stream's language

=head2 The Video::FFmpeg::AVStream::Audio class

=head3 bit_rate

the average bitrate in bit/s

=head3 sample_rate

samples per second (hz)

=head3 channels

number of audio channels 

=head2 The Video::FFmpeg::AVStream::Video class

=head3 width

picture width

=head3 height

picture height

=head3 fps

frames per second, 0 if not available. 

=head3 display_aspect

aspect ratio of the picture in "W:H" format

=head3 pixel_aspect

aspect ratio of the pixels in "W:H" format, or undef if not defined;

=head1 DESCRIPTION

Video::FFmpeg is a factory class for working with video files. Video::FFmpeg utilises FFmpeg's libavformat, and provides a basic interface.

=head2 EXPORT

None by default.

=head1 SEE ALSO

=over 4

=item L<Video::FFmpeg>

=item L<Video::FFmpeg::AVFormat>

=item L<Video::FFmpeg::AVStream>

=item L<Video::FFmpeg::AVStream::Audio>

=item L<Video::FFmpeg::AVStream::Video>

=item L<Video::FFmpeg::AVStream::Subtitle>

=item L<html://www.seattlenetworks.com/perl/FFmpeg>

=back

=head1 TODO

=head1 AUTHOR

Max Vohra, E<lt>max@seattlenetworks.comE<gt> L<html://www.seattlenetworks.com/>

This unauthorized version by Rene Schickbauer E<lt>rene.schickbauer@magnapowertrain.comE<gt>

=head UNAUTHORIZED VERSION

This unauthorized version fixes some critical bugs, see file "Changes" for details

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Max Vohra

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/Video/FFmpeg/AVStream.pm
package Video::FFmpeg::AVStream;

our $VERSION = '0.48';

1;
__END__

=head1 NAME

Video::FFmpeg::AVStream - Retrieve stream properties

=head1 SYNOPSIS

See Video::FFmpeg

=head1 METHODS

=head2 The Video::FFmpeg::AVStream class

=head3 codec

name of the codec

=head3 codec_type

returns one of "audio", "video", "subtitle", "data", "attachment", "data", or "unknown"

=head3 lang

returns the stream's language

=head2 The Video::FFmpeg::AVStream::Audio class

=head3 bit_rate

the average bitrate in bit/s

=head3 sample_rate

samples per second (hz)

=head3 channels

number of audio channels 

=head2 The Video::FFmpeg::AVStream::Video class

=head3 width

picture width

=head3 height

picture height

=head3 fps

frames per second, 0 if not available. 

=head3 display_aspect

aspect ratio of the picture in "W:H" format

=head3 pixel_aspect

aspect ratio of the pixels in "W:H" format, or undef if not defined;

=head2 The Video::FFmpeg::AVStream::Subtitle class

This class has no methods

=head1 DESCRIPTION

Video::FFmpeg is a factory class for working with video files. Video::FFmpeg utilises FFmpeg's libavformat, and provides a basic interface.

=head2 EXPORT

None by default.

=head1 SEE ALSO

=over 4

=item L<Video::FFmpeg>

=item L<Video::FFmpeg::AVFormat>

=item L<Video::FFmpeg::AVStream>

=item L<Video::FFmpeg::AVStream::Audio>

=item L<Video::FFmpeg::AVStream::Video>

=item L<Video::FFmpeg::AVStream::Subtitle>

=item L<html://www.seattlenetworks.com/perl/FFmpeg>

=back

=head1 TODO

=head1 AUTHOR

Max Vohra, E<lt>max@seattlenetworks.comE<gt> L<html://www.seattlenetworks.com/>

This unauthorized version by Rene Schickbauer E<lt>rene.schickbauer@magnapowertrain.comE<gt>

=head UNAUTHORIZED VERSION

This unauthorized version fixes some critical bugs, see file "Changes" for details

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Max Vohra

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/Video/FFmpeg/AVStream/Audio.pm
package Video::FFmpeg::AVStream::Audio;

our $VERSION = '0.48';

our @ISA = ('Video::FFmpeg::AVStream');

1;
__END__
=head1 NAME

Video::FFmpeg::AVStream::Audio - Retrieve properties of an audio stream

=head1 SYNOPSIS

  See Video::FFmpeg

=head1 METHODS

The Video::FFmpeg::AVStream::Audio class inherits methods from the Video::FFmpeg::AVStream class

=head2 The Video::FFmpeg::AVStream class

=head3 codec

name of the codec

=head3 codec_type

returns one of "audio", "video", "subtitle", "data", "attachment", "data", or "unknown"

=head3 lang

returns the stream's language

=head2 The Video::FFmpeg::AVStream::Audio class

=head3 bit_rate

the average bitrate in bit/s

=head3 sample_rate

samples per second (hz)

=head3 channels

number of audio channels 

=head1 DESCRIPTION

Video::FFmpeg is a factory class for working with video files. Video::FFmpeg utilises FFmpeg's libavformat, and provides a basic interface.

=head2 EXPORT

None by default.


=head1 SEE ALSO

=over 4

=item L<Video::FFmpeg>

=item L<Video::FFmpeg::AVFormat>

=item L<Video::FFmpeg::AVStream>

=item L<Video::FFmpeg::AVStream::Audio>

=item L<Video::FFmpeg::AVStream::Video>

=item L<Video::FFmpeg::AVStream::Subtitle>

=item L<html://www.seattlenetworks.com/perl/FFmpeg>

=back

=head1 TODO

=head1 AUTHOR

Max Vohra, E<lt>max@seattlenetworks.comE<gt> L<html://www.seattlenetworks.com/>

This unauthorized version by Rene Schickbauer E<lt>rene.schickbauer@magnapowertrain.comE<gt>

=head UNAUTHORIZED VERSION

This unauthorized version fixes some critical bugs, see file "Changes" for details

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Max Vohra

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/Video/FFmpeg/AVStream/Video.pm
package Video::FFmpeg::AVStream::Video;

our $VERSION = '0.48';

our @ISA = ('Video::FFmpeg::AVStream');

1;
__END__
=head1 NAME

Video::FFmpeg::AVStream::Video - Retrieve video properties common to all types of streams

=head1 SYNOPSIS

  See Video::FFmpeg

=head1 METHODS

The Video::FFmpeg::AVStream::Video

=head2 The Video::FFmpeg::AVStream class

=head3 codec

name of the codec

=head3 codec_type

returns one of "audio", "video", "subtitle", "data", "attachment", "data", or "unknown"

=head3 lang

returns the stream's language

=head2 The Video::FFmpeg::AVStream::Video class

=head3 width

picture width

=head3 height

picture height

=head3 fps

frames per second, 0 if not available. 

=head3 display_aspect

aspect ratio of the picture in "W:H" format

=head3 pixel_aspect

aspect ratio of the pixels in "W:H" format, or undef if not defined;

=head1 DESCRIPTION

Video::FFmpeg is a factory class for working with video files. Video::FFmpeg utilises FFmpeg's libavformat, and provides a basic interface.

=head2 EXPORT

None by default.


=head1 SEE ALSO

=over 4

=item L<Video::FFmpeg>

=item L<Video::FFmpeg::AVFormat>

=item L<Video::FFmpeg::AVStream>

=item L<Video::FFmpeg::AVStream::Audio>

=item L<Video::FFmpeg::AVStream::Video>

=item L<Video::FFmpeg::AVStream::Subtitle>

=item L<html://www.seattlenetworks.com/perl/FFmpeg>

=back

=head1 TODO

=head1 AUTHOR

Max Vohra, E<lt>max@seattlenetworks.comE<gt> L<html://www.seattlenetworks.com/>

This unauthorized version by Rene Schickbauer E<lt>rene.schickbauer@magnapowertrain.comE<gt>

=head UNAUTHORIZED VERSION

This unauthorized version fixes some critical bugs, see file "Changes" for details

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Max Vohra

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/Video/FFmpeg/AVStream/Subtitle.pm
package Video::FFmpeg::AVStream::Subtitle;

our $VERSION = '0.48';

our @ISA = ('Video::FFmpeg::AVStream');

1;

__END__
=head1 NAME

Video::FFmpeg::AVStream::Subtitle - Retrieve video properties common to all types of streams

=head1 SYNOPSIS

  See Video::FFmpeg

=head1 METHODS

The Video::FFmpeg::AVStream::Video

=head2 The Video::FFmpeg::AVStream class

=head3 codec

name of the codec

=head3 codec_type

returns one of "audio", "video", "subtitle", "data", "attachment", "data", or "unknown"

=head3 lang

returns the stream's language

=head2 The Video::FFmpeg::AVStream::Subtitle class

This class has no methods

=head1 DESCRIPTION

Video::FFmpeg is a factory class for working with video files. Video::FFmpeg utilises FFmpeg's libavformat, and provides a basic interface.

=head2 EXPORT

None by default.


=head1 SEE ALSO

=over 4

=item L<Video::FFmpeg>

=item L<Video::FFmpeg::AVFormat>

=item L<Video::FFmpeg::AVStream>

=item L<Video::FFmpeg::AVStream::Audio>

=item L<Video::FFmpeg::AVStream::Video>

=item L<Video::FFmpeg::AVStream::Subtitle>

=item L<html://www.seattlenetworks.com/perl/FFmpeg>

=back

=head1 TODO

=head1 AUTHOR

Max Vohra, E<lt>max@seattlenetworks.comE<gt> L<html://www.seattlenetworks.com/>

This unauthorized version by Rene Schickbauer E<lt>rene.schickbauer@magnapowertrain.comE<gt>

=head UNAUTHORIZED VERSION

This unauthorized version fixes some critical bugs, see file "Changes" for details

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Max Vohra

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/Video/FFmpeg.pm
package Video::FFmpeg;

use 5.010000;
use strict;
use warnings;

use Video::FFmpeg::AVFormat;
use Video::FFmpeg::AVStream;
use Video::FFmpeg::AVStream::Video;
use Video::FFmpeg::AVStream::Audio;
use Video::FFmpeg::AVStream::Subtitle;

our @ISA = ('Video::FFmpeg::AVFormat');

our $VERSION = '0.48';

sub new {
	my $class = shift;
	die "File $_[0] doesn't exist!" unless(-f $_[0]);
	my $avfrmt = Video::FFmpeg::AVFormat::open($_[0]);
		
	bless $avfrmt, $class;
	return $avfrmt;
};

require XSLoader;
XSLoader::load('Video::FFmpeg', $VERSION);

1;
__END__

=head1 NAME

Video::FFmpeg - Retrieve video properties using libavformat such as: height width codec fps

=head1 SYNOPSIS

  use 5.010;
  use Video::FFmpeg;

  my $info = Video::FFmpeg::AVFormat->new($ARGV[0]);

  print "Duration: ",$info->duration,"\n";

  my @video = $info->video;
  print "num video streams: ",$#video+1,"\n";
  my @audio = $info->audio;
  print "num audio streams: ",$#audio+1,"\n";
  my @sub = $info->subtitle;
  print "num sub streams: ",$#sub+1,"\n";

  my @streams = $info->streams;
  for my $id (0 .. $#streams){
    my $stream = $streams[$id];
    print $stream->codec_type,"stream $id\n";
    print "\ttype: ",$stream->codec_type,"\n";
    print "\tcodec: ",$stream->codec,"\n";
    print "\tlanguage: ",$stream->lang,"\n";
    given($stream->codec_type){
      when("video") {
        print "\tfps: ",$stream->fps,"\n";
        print "\tDAR: ",$stream->display_aspect,"\n";
      }
      when("audio") {
        print "\tsample rate: ",$stream->sample_rate,"hz\n";
        print "\taudio language: ",$stream->lang,"\n";
      }
      when("subtitle") {
        print "\tsub codec: ",$stream->codec,"\n";
        print "\tsub language: ",$stream->lang,"\n";
      }
    }
  };

=head1 METHODS

=head2 The Video::FFmpeg::AVFormat class

=head3 Video::FFmpeg::AVFormat->new($file)

AVFormat Constructor

=head3 filename

returns the filename of the $info object

=head3 duration

duration of the stream, in HH:MM:SS.MS. 

=head3 start_time

position of the first frame of the component, in microseconds

=head3 bit_rate

total stream bitrate in bit/s, 0 if not available. 

=head3 video

if called in scalar context, returns the first Video::FFmpeg::AVStream::Video object. if called in list context, it returns all Video::FFmpeg::AVStream::Video objects

=head3 audio

if called in scalar context, returns the first Video::FFmpeg::AVStream::Audio object. if called in list context, it returns all Video::FFmpeg::AVStream::Audio objects

=head3 subtitles

if called in scalar context, returns the first Video::FFmpeg::AVStream::Subtitle object. if called in list context, it returns all Video::FFmpeg::AVStream::Subtitle objects

=head3 streams

returns a list of all Video::FFmpeg::AVStream objects.


=head2 The Video::FFmpeg::AVStream class

=head3 codec

name of the codec

=head3 codec_type

returns one of "audio", "video", "subtitle", "data", "attachment", "data", or "unknown"

=head3 lang

returns the stream's language

=head2 The Video::FFmpeg::AVStream::Audio class

=head3 bit_rate

the average bitrate in bit/s

=head3 sample_rate

samples per second (hz)

=head3 channels

number of audio channels 

=head2 The Video::FFmpeg::AVStream::Video class

=head3 width

picture width

=head3 height

picture height

=head3 fps

frames per second, 0 if not available. 

=head3 display_aspect

aspect ratio of the picture in "W:H" format

=head3 pixel_aspect

aspect ratio of the pixels in "W:H" format, or undef if not defined;

=head1 DESCRIPTION

Video::FFmpeg is a factory class for working with video files. Video::FFmpeg utilises FFmpeg's libavformat, and provides a basic interface.

=head2 EXPORT

None by default.

=head1 SEE ALSO

=over 4

=item L<Video::FFmpeg>

=item L<Video::FFmpeg::AVFormat>

=item L<Video::FFmpeg::AVStream>

=item L<Video::FFmpeg::AVStream::Audio>

=item L<Video::FFmpeg::AVStream::Video>

=item L<Video::FFmpeg::AVStream::Subtitle>

=item L<html://www.seattlenetworks.com/perl/FFmpeg>

=back

=head1 TODO

=head1 AUTHOR

Max Vohra, E<lt>max@seattlenetworks.comE<gt> L<html://www.seattlenetworks.com/>

This unauthorized version by Rene Schickbauer E<lt>rene.schickbauer@magnapowertrain.comE<gt>

=head UNAUTHORIZED VERSION

This unauthorized version fixes some critical bugs, see file "Changes" for details

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Max Vohra

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut
